### PR TITLE
ejabberd: refactor module, add test

### PR DESCRIPTION
**Pull request by ajs124**
###### Motivation for this change
I run ejabberd.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by dotlambda** [comment]
Can't we use [`tmpfiles`](https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html#) for e.g. creating `/var/lock/ejabberdctl` and then run the complete unit as the correct user? See also https://github.com/NixOS/nixpkgs/pull/56265.

File: nixos/tests/ejabberd.nix [hunk 1]
```diff
@@ -0,0 +1,303 @@
+import ./make-test.nix {
+  name = "ejabberd";
+  nodes = {
+    client = { nodes, pkgs, ... }: {
+      environment.systemPackages = let
+        sendMessage = pkgs.writeScriptBin "send-message" ''
+          #!/usr/bin/env python3
```
Comment (dotlambda) [inline nixos/tests/ejabberd.nix]:
```suggestion
          #! ${(pkgs.python3.withPackages (ps: [ ps.sleekxmpp ])).interpreter}
```

File: nixos/tests/ejabberd.nix [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+              # If you want to verify the SSL certificates offered by a server:
+              # xmpp.ca_certs = "path/to/ca/cert"
+
+              if xmpp.connect(('${nodes.server.config.networking.primaryIPAddress}', 5222)):
+                  xmpp.process(block=True)
+              else:
+                  print("Unable to connect.")
+                  sys.exit(1)
+        '';
+      in [ (pkgs.python3.withPackages (ps: [ ps.sleekxmpp ])) sendMessage ];
```
Comment (dotlambda) [inline nixos/tests/ejabberd.nix]:
```suggestion
      in [ sendMessage ];
```

File: nixos/tests/ejabberd.nix [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      services.ejabberd = {
+        enable = true;
+        configFile = "/etc/ejabberd.yml";
+      };
+
+      environment.etc."ejabberd.yml" = {
+        user = "ejabberd";
+        mode = "0600";
+        text = ''
+loglevel: 3
```
Comment (dotlambda) [inline nixos/tests/ejabberd.nix]: This should be indented. The indentation will automatically be removed as explained in https://nixos.org/nix/manual/#idm140737317975776.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: nixos/tests/ejabberd.nix (hunk 1 above)
Comment (ajs124) [reply nixos/tests/ejabberd.nix]: This python script is lifted from the prosody test, so this should probably be fixed there, as well.

Comment (dotlambda) [reply nixos/tests/ejabberd.nix]: It's not a necessity because both ways work fine. However, it would be nice to have a single file for the script.

**Comment by aanderse** [comment]
> Can't we use [`tmpfiles`](https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html#) for e.g. creating `/var/lock/ejabberdctl` and then run the complete unit as the correct user? See also #56265.

A big +1 on not using escalated privileges to launch this service.

**Comment by ajs124** [comment]
Seems like `/var/lock/ejabberdctl` is not needed for anything anymore, so I dropped that and converted the other two directories to `tmpfiles`.

File: nixos/modules/services/networking/ejabberd.nix [hunk 4]
```diff
@@ -149,13 +138,18 @@ in {
...
+              ${ectl} load "$dump"
             done
           done
         fi
       '';
     };
 
+    systemd.tmpfiles.rules = [
+      "d ${cfg.logsDir} 0750 ${cfg.user} ${cfg.group} -"
+      "d ${cfg.spoolDir} 0750 ${cfg.user} ${cfg.group} -"
```
Comment (aanderse) [inline nixos/modules/services/networking/ejabberd.nix]: Please escape `${cfg.logsDir}` and `${cfg.spoolDir}` with single quotes.

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/tests/ejabberd.nix (hunk 1 above)
Comment (ajs124) [reply nixos/tests/ejabberd.nix]: How about something like this?

File: nixos/modules/services/networking/ejabberd.nix (hunk 4 above)
Comment (ajs124) [reply nixos/modules/services/networking/ejabberd.nix]: Like [this](https://github.com/NixOS/nixpkgs/pull/59731/commits/f9406ddc080209ef38f9bcaf018c25461bbdebae#diff-f4ad427ecde914245907dd54f8d02c4dR149)?

Comment (aanderse) [reply nixos/modules/services/networking/ejabberd.nix]: Yes. Though I now notice you're missing the final parameter. Please add one more `-` to each entry.

Comment (ajs124) [reply nixos/modules/services/networking/ejabberd.nix]: I am? Neither the examples in the manual, e.g. 
```
# /usr/lib/tmpfiles.d/abrt.conf
d /var/tmp/abrt 0755 abrt abrt -
```
nor the examples in nixpkgs, e.g. [iwd](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/iwd.nix#L28) have it.

Comment (aanderse) [reply nixos/modules/services/networking/ejabberd.nix]: Ah you're totally right. Even the [manual](https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html) has an example like that. Silly me. Sorry :smile:

File: nixos/tests/ejabberd.nix [hunk 5]
```diff
@@ -0,0 +1,259 @@
+import ./make-test.nix {
```
Comment (aanderse) [inline nixos/tests/ejabberd.nix]: Would you please list yourself as the maintainer of this test?

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/tests/ejabberd.nix (hunk 5 above)
Comment (ajs124) [reply nixos/tests/ejabberd.nix]: Done.

**Comment by aanderse** [comment]
@GrahamcOfBorg test ejabberd prosody

File: nixos/tests/ejabberd.nix [hunk 6]
```diff
@@ -0,0 +1,262 @@
+import ./make-test.nix ({ pkgs, ... }: {
```
Comment (aanderse) [inline nixos/tests/ejabberd.nix]: Please add this to all-tests.nix

Comment (ajs124) [reply nixos/tests/ejabberd.nix]: Thanks for walking me through this… I added it now.

**Comment by ajs124** [comment]
How does that even happen? It worked on aarch64 but not x86_64? It works fine on my laptop and desktop.

**Comment by aanderse** [comment]
> How does that even happen? It worked on aarch64 but not x86_64? It works fine on my laptop and desktop.

`Cookie file /var/lib/ejabberd/.erlang.cookie must be accessible by owner only`
Sounds like your `spoolDir` needs to be created as 0700 and enforced.

**Comment by ajs124** [comment]
It's worth a try. I don't know if it makes a difference though, since it was already working for me on my machines.

I mean, how and why would the mode of the directory influence the mode of its contents?

**Comment by aanderse** [comment]
It was just a guess based on the error message. With the current mode of the directory the file is accessible to more than just the owner.

I'm also confused by this error because it worked on my machine as well.

**Review by aanderse — APPROVED** [review]

**Comment by aanderse** [comment]
@ajs124 Fantastic work! :tada: 
@dotlambda I'm very happy with this PR. Anything outstanding from your perspective?

**Review by dotlambda — APPROVED** [review]
Very nice! I did not test it though since I do not run ejabberd yet.
You might want to change the last commit message to use `nixos/ejabberd`.

**Comment by Infinisil** [comment]
@GrahamcOfBorg test ejabberd
@GrahamcOfBorg test prosody

**Pull request merged**